### PR TITLE
Add meson-based build

### PR DESCRIPTION
**Pull request by grlee77**
## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Comment by grlee77** [comment]
For context, @stefanv, @matthew-brett and myself had a couple hours at the NumFOCUS project summit to try this out. We were able to successfully build a small subset of the modules, but still need to add/update the `meson.build` files for the remaining modules.

**Comment by stefanv** [comment]
Ran the tests successfully locally.  Still need to update CI to use meson.

**Comment by lagru** [comment]
Thanks for exploring this! For those new to this idea, [How to build SciPy with Meson](https://scipy.github.io/devdocs/dev/contributor/meson.html) answered a lot of my early questions around this, especially the FAQ. For me it is worth noting, that `meson` seems to be an easy build dependency to supply because it's available on PyPI and conda-forge as a Python package.

After testing the build system with SciPy, the one thing I currently really don't like about the separation of the source and installation location, is that is breaks the debugging workflow for my IDE; PyCharm in my case but I expect this to be true for other IDEs as well. The files that the IDE picks up as being executed aren't recognized as the currently edited source. This makes debugging, setting breakpoints and some other tools a lot more inconvenient. :/ Apart from that this seems like an improvement in a lot of places. (Putting this here, because I don't think we have an official place to discuss this yet.)

**Comment by grlee77** [comment]
> After testing the build system with SciPy, the one thing I currently really don't like about the separation of the source and installation location, is that is breaks the debugging workflow for my IDE; PyCharm in my case but I expect this to be true for other IDEs as well. The files that the IDE picks up as being executed aren't recognized as the currently edited source. This makes debugging, setting breakpoints and some other tools a lot more inconvenient. :/ Apart from that this seems like an improvement in a lot of places. (Putting this here, because I don't think we have an official place to discuss this yet.)

I just opened #6540 if you want to discuss things there. If it is Meson-specific it also seems suitable within this PR. I did see some discussion of this IDE concern in the "How to Build SciPy with Meson" [FAQ section](https://scipy.github.io/devdocs/dev/contributor/meson.html#meson-frequently-asked-questions), but not for PyCharm specifically. I suspect better handling and best practices around this will become more common over the next year, but don't know of specific work to point to.

**Comment by stefanv** [comment]
Failures unrelated; we have warnings in our tests due to usage of `np.int` in scipy (deprecated for using `int` by itself).

**Comment by stefanv** [comment]
have to add ninja

**Comment by stefanv** [comment]
CircleCI doc build failing; likely running out of memory due to parallel builds.

**Comment by stefanv** [comment]
:heavy_check_mark: Green!

File: skimage/_build_utils/system_info.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                else:
+                    warnings.warn(BlasNotFoundError.__doc__, stacklevel=2)
+                    blas_src_info = get_info('blas_src')
+                    if not blas_src_info:
+                        warnings.warn(BlasSrcNotFoundError.__doc__, stacklevel=2)
+                        return
+                    dict_append(info, libraries=[('fblas_src', blas_src_info)])
+
+            self.set_info(**info)
+            return
```
Comment (jarrodmillman) [inline skimage/_build_utils/system_info.py:211]:
```suggestion
# For new enough numpy.distutils, the ACCELERATE=None environment
# variable in the top-level setup.py is enough, so no need to
# customize BLAS detection.
get_info = old_get_info
```

**Review by jarrodmillman — CHANGES_REQUESTED** [review]

**Comment by stefanv** [comment]
@grlee77 @lagru I think this is pretty much ready; want to take a look?

**Review by lagru — COMMENTED** [review]
I'm not familiar with meson or similar systems at all so, my current feedback may quite superficial still. :wink:

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -85,10 +85,10 @@ jobs:
           set -ex
           source tools/github/before_install.sh
           set -ex
-          python setup.py sdist
           if [[ $INSTALL_FROM_SDIST == "1" ]]; then
-              pip uninstall cython -y;
-              pip install dist/scikit-image-*.tar.gz;
+              pip uninstall cython -y
```
[the remark this answers is not in the document]
Comment (stefanv) [reply .github/workflows/tests.yml:89]: This is to ensure that the compiled C files are included in the sdist. If they're missing and Cython is present, our build system will generate them.

File: pyproject.toml [hunk 3]
```diff
@@ -1,5 +1,52 @@
...
+    {name = "scikit-image developers", email="skimage-core@discuss.scientific-python.org"},
+]
+requires-python = ">=3.8"
+dependencies = [
+    # TODO: update to "pin-compatible" once possible, see
+    # https://github.com/FFY00/meson-python/issues/29
+    "numpy>=1.20",
+]
+readme = "README.md"
+classifiers = [
```
[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:15]: Yes, for now we support the two build systems in parallel, but eventually we'll just have pyproject.toml.

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,52 @@
...
+    'Operating System :: MacOS',
+]
+dynamic = ['version']
+
+[project.urls]
+homepage = "https://scikit-image.org"
+documentation = "https://scikit-image.org/docs/stable"
+source = "https://github.com/scikit-image/scikit-image"
+download = "https://pypi.org/project/scikit-image/#files"
+tracker = "https://github.com/scikit-image/scikit-image/issues"
```
[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:42]: Not sure what you're referring to; we don't usually capitalize URLs.

File: pyproject.toml [hunk 5]
```diff
@@ -1,5 +1,52 @@
...
+
 [build-system]
+build-backend = 'mesonpy'
 requires = [
+    "meson-python>=0.10.0",
```
[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:47]: `meson-python` pulls in ninja, so I've removed it from the `requirements/build.txt`

File: meson.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# Building with Meson
+
+### Developer build
+
+**Install build tools:** `pip install meson ninja cython pythran`
+
+**Generate ninja make files:** `meson build --prefix=$PWD/build`
+
+**Compile:** `ninja -C build`
```
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:9]: I am being explicit here about what is happening behind the scenes. `python -m build --no-isolation` does the same thing.  I will add that as shorthand.

File: pyproject.toml [hunk 7]
```diff
@@ -1,5 +1,52 @@
...
+    # `wheel` is needed for non-isolated builds, given that `meson-python`
+    # doesn't list it as a runtime requirement (at least in 0.5.0)
```
[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:49]: Fixed

File: INSTALL.rst [hunk 8]
```diff
@@ -300,11 +300,11 @@ When using ``venv``, you may find the following bash commands useful:
   # Make sure that pip is up to date
   pip install --upgrade pip
   # Install all development and runtime dependencies of scikit-image
-  pip install -r <(cat requirements/*.txt)
+  pip install -r requirements.txt
   # Build and install scikit-image from source
-  pip install -e . -vv
+  pip install -e . -vv  ## TODO: to be updated for meson (see meson.md)
```
[the remark this answers is not in the document]
Comment (stefanv) [reply INSTALL.rst:305]: meson does not do in-place builds

File: meson.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# Building with Meson
+
+### Developer build
+
+**Install build tools:** `pip install meson ninja cython pythran`
+
+**Generate ninja make files:** `meson build --prefix=$PWD/build`
```
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:7]: Done

[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:7]: I can't reproduce this. I fixed the `pip install` instruction to install `meson-py` instead.

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply pyproject.toml:42]: I assume this refers to the left hand side labels.

[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:42]: Looks like these are predefined for `pyproject.toml`, e.g. https://python-poetry.org/docs/pyproject/#homepage

File: INSTALL.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply INSTALL.rst:305]: This is using a legacy setuptools editable install. [PEP 660](https://peps.python.org/pep-0660/) editable installs are not yet implemented for the `meson-python` build backend.

File: meson.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+**Compile:** `ninja -C build`
+
+**Install:** `meson install -C build`
+
+The install step copies the necessary Python files into the build dir to form a complete package.
```
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.md:13]: PEP 660 editable installs would probably solve this; see https://github.com/FFY00/meson-python/issues/47 for the relevant tracking issue.

File: meson.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.md:7]: https://github.com/mesonbuild/meson/pull/10445 updates this error message.

```
Note that this error message can also be triggered by
external dependencies being installed within your source
tree - it's not recommended to do this.
```

tl;dr you probably have numpy installed in a venv that is a subdirectory of your scikit-image clone. This confuses Meson as it doesn't know whether it's supposed to be a project source or an external dependency.

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply pyproject.toml:42]: That is talking about the `[tool.poetry]` section, which is only relevant to poetry.

Currently, scikit-image has all these labels, but they are defined here:
https://github.com/scikit-image/scikit-image/blob/c1af2bf80d2c8845bbbf06222f87e8b58a8e5c1e/setup.py#L26-L33

https://github.com/scikit-image/scikit-image/blob/c1af2bf80d2c8845bbbf06222f87e8b58a8e5c1e/setup.py#L221-L224

setuptools natively handles url= and download_url=, and the rest are all capitalized in the dict that is passed to setuptools. The resulting wheel metadata looks like this:

```
Home-page: https://scikit-image.org
Download-URL: https://scikit-image.org/docs/stable/install.html
Project-URL: Bug Tracker, https://github.com/scikit-image/scikit-image/issues
Project-URL: Documentation, https://scikit-image.org/docs/stable/
Project-URL: Source Code, https://github.com/scikit-image/scikit-image
```

PEP 621 doesn't really seem to say whose responsibility it is to capitalize them.

[the remark this answers is not in the document]
Comment (stefanv) [reply pyproject.toml:42]: Looks like `homepage` and `download` are special: if you capitalize them, they also become Project-URL. PEP 621 gives the following example:

```
[project.urls]
homepage = "https://example.com"
documentation = "https://readthedocs.org"
repository = "https://github.com"
changelog = "https://github.com/me/spam/blob/master/CHANGELOG.md"
```

File: meson.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:13]: @lagru On the other hand, meson & ninja build skimage in under a minute on my machine! There will always be quirks when moving to a new build system but, after working with it for a bit, I am confident that Meson is the right tool for the job.

[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:13]: @eli-schwartz I wonder if that issue should be closed and a new one opened with the proposed plan. Right now, almost the whole conversation there is moot.

File: meson.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+Do not skip this step, or the package won't work.
+
+To use the package, add it to your PYTHONPATH:
+
+```
+export PYTHONPATH=${PWD}/build/lib64/python3.10/site-packages
+pytest --pyargs skimage
+```
```
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:21]: As soon as I add a dev.py, I'll add this in.

File: meson.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:13]: It looks like the shortest path to success is to [have meson install symlinks](https://github.com/FFY00/meson-python/issues/47#issuecomment-1101539767) and then to point to the build directory in the package. This should be very achievable; do you know if anything is blocking this, or whether I should look into implementing it?

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.md:13]: AFAIK nothing is blocking meson-python from implementing editable installs... simply that no one got around to implementing it.

However @FFY00 would know more.

File: .github/workflows/tests.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (lagru) [reply .github/workflows/tests.yml:89]: Ah, got it. Without Cython we ensure the CI will fails if the C sources are not included.

File: meson.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (lagru) [reply meson.md:7]: Ah, after using a venv outside the project dir and moving everything to a filesystem not on tmpfs the error disappeared. Thanks! That's probably something we should document once we recommend this approach in our contribution guide.

File: INSTALL.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (lagru) [reply INSTALL.rst:305]: Is this legacy install triggered through meson? Is that why meson is shown in the installations output?

File: meson.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lagru) [reply meson.md:13]: Yeah, building *is* a lot faster with ninja. I'm not vetoing this anyway because in the long-term we have to find an alternative to distutils. :shrug: 

Though it's good to know that there is an intention to support the editable install workflow. From SciPy's meson guide that wasn't clear. Thanks @eli-schwartz, you feedback has been very helpful!

File: meson.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lagru) [reply meson.md:9]: Good to know. Though testing locally I'm notified that it needs `patchelf` to build the wheel. After installing that, `build`, `wheel` and the appropriate NumPy version makes `python -m build --no-isolation` work. But we can look at that once we get there.

File: meson.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply meson.md:13]: @eli-schwartz Until we get editable builds installed, do you know if there is a way to specify in `pyproject.toml` that you want the default build system to be meson, but that you want editable installs to go via setuptools.  Obviously not a long term solution, but could be a band-aid for @lagru's use-case.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.md:13]: I don't think pyproject.toml has any support whatsoever for allowing two build systems, there's no thought given to things like "perhaps we want to switch build systems and leave the old one as a fallback during a transition period".

You can invoke the setuptools one with `python setup.py develop`, which editable installs use under the hood for setuptools (this is why a PEP was needed to define what an editable install actually is).

@rgommers can confirm this but iirc SciPy "solved" it by having two pyproject.toml files and copying the backup one in place if you wanted to switch build systems.

[the remark this answers is not in the document]
Comment (rgommers) [reply meson.md:13]: > You can invoke the setuptools one with `python setup.py develop`,

Yep, that's the right solution for as long as you want to keep `setup.py`'s around

> @rgommers can confirm this but iirc SciPy "solved" it by having two pyproject.toml files and copying the backup one in place if you wanted to switch build systems.

Almost, but not exactly. We recommend packagers who want to do this to simply delete the one `build-backend` line in `pyproject.toml`. That way the build falls back to the default (= setuptools. 

There's a significant flaw in Pip's transition plan for PEP 517 users. I thought we could ship two build systems, and do `pip install --no-use-pep517`, but that doesn't work (for <reasons>). So you can have two build systems in your source tree, but without a patch you can only use one through Pip.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.md:13]: Thanks for clarifying. :) It was more or less like I remembered. I suppose editing the pyproject.toml file vs. copying a different one in place isn't all that different.

**Comment by stefanv** [comment]
@lagru Is the not-in-place builds a blocker for you?  If not, is there anything else that I need to do to get this in.  I can add a `dev.py` hook if that'd be helpful.

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -85,10 +85,10 @@ jobs:
...
+              python setup.py sdist
+              pip install dist/scikit-image-*.tar.gz
           else
               pip install -vv -e .;
```
Comment (lagru) [inline .github/workflows/tests.yml:93]: So is our CI actually testing the new workflow and build system in `meson.md`? As mentioned in https://github.com/scikit-image/scikit-image/pull/6536#discussion_r995193616, this is actually using setuptools under the hood. And the `setup.py sdist` approach doesn't use meson either.

Are you waiting with CI integration until `dev.py` is added?

File: meson.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+**Install:** `meson install -C build`
+
+The install step copies the necessary Python files into the build dir to form a complete package.
+Do not skip this step, or the package won't work.
+
+To use the package, add it to your PYTHONPATH:
+
+```
+export PYTHONPATH=${PWD}/build/lib64/python3.10/site-packages
```
Comment (lagru) [inline meson.md:19]: On my system this path is different: `build/lib/python3.10/site-packages`...?

File: skimage/_shared/meson.build [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+python_sources = [
+  '__init__.py',
+  '_dependency_checks.py',
+  '_geometry.py',
+  '_tempfile.py',
+  '_warnings.py',
+  'coord.py',
+  'filters.py',
+  'lazy.py',
```
Comment (lagru) [inline skimage/_shared/meson.build:41]:
```suggestion
```
I think the CI fails because it test's on the would-be merged state (what's the proper name?). `lazy.py` was removed in #6577 so it's not found and causes the CI to fail.

**Review by lagru — COMMENTED** [review]
To be honest, I don't think I understand the full build system (yet). Testing it locally it seem to do what is expected though and the tests pass both for the `ninja -C build` approach and for `pip install --no-build-isolation .`. Therefore, I'm happy to merge this if my other comments are addressed. Sorry for holding this up a bit longer.

**Comment by stefanv** [comment]
> One comment that may be helpful: there's an `install_subdir` function in Meson that may be appropriate to use in case you have a directory with pure Python files that you all want to install. In SciPy I've preferred adding `meson.build` files to every sub-directory, however Pandas is going with `install_subdir` - see in `pandas/meson.build` at [pandas-dev/pandas#49115](https://github.com/pandas-dev/pandas/pull/49115).

Thanks, Ralf! I think we indeed have a few directories where we just need a vanilla install, so I may use that there.

File: meson.md (hunk 13 above)
Comment (stefanv) [reply meson.md:19]: Yes, the `dev.py` tool takes this into consideration, but these notes obviously cannot cover all platform intricacies.

File: .github/workflows/tests.yml (hunk 12 above)
Comment (stefanv) [reply .github/workflows/tests.yml:93]: No, if you look at the build logs, you'll see meson is used in various places, since the build system in pyproject.toml has been changed to Meson.

Comment (lagru) [reply .github/workflows/tests.yml:93]: Yes, I can see that meson is used. However, as mentioned in the linked comment [`pip install -e .` (in-place install)](https://github.com/scikit-image/scikit-image/actions/runs/3292132986/jobs/5427133838#step:4:146) uses setuptools and when I tested this locally there was no parallel build happening for this command. Just the slow consecutive build with setuptools. I'm asking whether we should also test the explicit workflow that doesn't use a in-place install, e.g. invoking `ninja`. And take advantage of the parallel feature in our CI. 

Maybe I'm just misunderstanding something; this interaction between the different tools and build systems still seems very opaque to me.

**Comment by lagru** [comment]
Not sure what the [CI failure](https://github.com/scikit-image/scikit-image/actions/runs/3292132986/jobs/5427134608#step:5:275) is about. It seems like the missing `skimage/__init__.pyi` is not included in the sdist archive?

**Comment by stefanv** [comment]
Yes, that's one of the annoyances with meson: everything is specified by hand.  Maybe we'll automate some of that later.

**Review by lagru — APPROVED** [review]

File: .github/workflows/tests.yml (hunk 12 above)
Comment (lagru) [reply .github/workflows/tests.yml]: Discussed in the community meeting: the CI test's a non-editable install from an sdist. Also it seems that `pip install -e .` does both, building with ninja *and* building with setuptools. So either way the CI is testing the meson build.

**Comment by stefanv** [comment]
The no-pooch test target fails because, correctly, it cannot find `no_time_for_that_tiny.gif`

**Comment by stefanv** [comment]
@lagru @grlee77 Green! :heavy_check_mark:

**Review by lagru — APPROVED** [review]
I added `pip install --no-build-isolation .` in two more places and made skipping the pooch reliant tests more specific. Otherwise looks good. Excited to merge this, once the CI is green!

**Comment by lagru** [comment]
Thanks everyone for the effort! :tada:

**Comment by stefanv** [comment]
Yeah!!! 🎸⚡ Next up, I will add dev.py.
